### PR TITLE
Allow `card_` prefix when validating payment method IDs to fix failing subscription renewals

### LINKED ISSUE #6490: Subscription renewals fail if payment method uses `card_` prefix
### Describe the bug

WC Subscription renewals fail with the error `card_xxx is not a valid Stripe identifier` if the payment method id is saved with a `card_` prefix. The `card_` IDs were issued under a legacy Stripe API that we no longer use. However, these are accepted as a valid payment method when renewing a subscription.

We currently apply validation for stripe via the [`WCPay\Core\Server\Request::validate_stripe_id()`](https://github.com/Automattic/woocommerce-payments/blob/1356f2f6101e49497f06eb9e0eff1eafb37236d5/includes/core/server/class-request.php#L494) method.

2 calls to `validate_stripe_response()` that enforce `'pm', 'src'`:
* [`WCPay\Core\Server\Request\Create_Intention::set_payment_method`](https://github.com/Automattic/woocommerce-payments/blob/1356f2f6101e49497f06eb9e0eff1eafb37236d5/includes/core/server/request/class-create-intention.php#L50)
* [`WCPay\Core\Server\Request\Create_Setup_Intention::set_payment_method`](https://github.com/Automattic/woocommerce-payments/blob/1356f2f6101e49497f06eb9e0eff1eafb37236d5/includes/core/server/request/class-create-and-confirm-setup-intention.php#L81)

Calling these with `[ 'pm', 'src', 'card' ]` has fixed the immediate issue on WCCOM: p1686268201180419-slack-CGGCLBN58

### To Reproduce

This is not straightforward to reproduce, since Stripe no longer issues `card_` payment method IDs.

### Actual behavior

Subscription renewals fail for subscriptions with an associated `card_` payment method.

### Expected behavior

Subscription renewals should process successfully for subscriptions with an associated `card_` payment method.

### Additional context

* Slack discussion: p1686268201180419-slack-CGGCLBN58
* We're reasonably confident that WCCOM is the only store impacted by this bug, since it only applies to stores that had payment methods migrated from a non-WCPay source

### PR DESCRIPTION
**Pull request by Jinksi**
Fixes #6490

#### Changes proposed in this Pull Request

This PR allows the `card_` prefix when validating payment method IDs. This fixes failing subscription renewals for stores that have stored legacy `card_` payment method IDs: namely WCCOM.

2 calls to `validate_stripe_response()` that enforce `'pm', 'src'` now allow the prefix `'card'`:
* [`WCPay\Core\Server\Request\Create_Intention::set_payment_method`](https://github.com/Automattic/woocommerce-payments/blob/1356f2f6101e49497f06eb9e0eff1eafb37236d5/includes/core/server/request/class-create-intention.php#L50)
* [`WCPay\Core\Server\Request\Create_Setup_Intention::set_payment_method`](https://github.com/Automattic/woocommerce-payments/blob/1356f2f6101e49497f06eb9e0eff1eafb37236d5/includes/core/server/request/class-create-and-confirm-setup-intention.php#L81)

#### Testing instructions

**Note:** This original issue described in #6490 is not straightforward to reproduce, since Stripe no longer issues `card_` payment method IDs.

**Regression testing**
1. Install WooCommerce Subscriptions alongside WCPay.
1. Purchase a subscription product. Note the subscription ID as shopper.
2. Navigate to this subscription via **WC→ Subscriptions** as merchant.
3. Process a WC Subscription renewal using the Actions dropdown. You may have to run the scheduled action `woocommerce_scheduled_subscription_payment` in WC → Status → Scheduled Actions.
4. Confirm the renewal order is processed successfully.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply) **N/A**

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6491` or branch name `fix/6490-allow-payment-method-card-prefix` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/6490-allow-payment-method-card-prefix) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 80b764120391e68b1a28608574f44c3de2ebdfe9
- Build time: 2023-06-14 09:08:18 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Review by hesahesa — APPROVED** [review]
we verified the same approach in WCCOM and it works

**Comment by Jinksi** [comment]
@Automattic/gamma I've requested your review on this due to your expertise around the `validate_stripe_id()` introduced in [WCPay as a Platform: Phase 1](https://github.com/Automattic/woocommerce-payments/pull/5321).

Is there any risk introduced here if we allow `card_` prefix for payment methods?

File: includes/core/server/request/class-create-intention.php [hunk 1]
```diff
@@ -47,7 +47,7 @@ public function get_method(): string {
 	 * @throws Invalid_Request_Parameter_Exception
 	 */
 	public function set_payment_method( string $payment_method_id ) {
-		$this->validate_stripe_id( $payment_method_id, [ 'pm', 'src' ] );
+		$this->validate_stripe_id( $payment_method_id, [ 'pm', 'src', 'card' ] );
```
Comment (htdat) [inline includes/core/server/request/class-create-intention.php:51]: Not urgent but I think we might want to add more tests to cover for this change: 

https://github.com/Automattic/woocommerce-payments/blob/96a1efd7b9e94491a34c0f929661d9c8f636e353/tests/unit/core/server/request/test-class-core-create-and-confirm-intention-request.php#L79-L83

Also, this method is being used in the child class `Create_And_Confirm_Intention` so we might lift it up to this class?

File: includes/core/server/request/class-create-and-confirm-setup-intention.php [hunk 2]
```diff
@@ -79,7 +79,7 @@ public function set_metadata( array $metadata ) {
 	 * @throws Invalid_Request_Parameter_Exception
 	 */
 	public function set_payment_method( string $payment_method_id ) {
-		$this->validate_stripe_id( $payment_method_id, [ 'pm', 'src' ] );
+		$this->validate_stripe_id( $payment_method_id, [ 'pm', 'src', 'card' ] );
```
Comment (htdat) [inline includes/core/server/request/class-create-and-confirm-setup-intention.php:83]: @Automattic/gamma - it sounds like this request class `Create_And_Confirm_Setup_Intention` does not have tests under `tests/unit/core/server/request`. I think we will need to add tests for it.

**Review by htdat — COMMENTED** [review]
> Is there any risk introduced here if we allow `card_` prefix for payment methods?

AFAIK, there is no significant risk. If needed for an urgent patch, yes, we can use it. 

Other than that, I have some inline comments, mostly in regard to PHP Unit tests. 

**_Update_: it would be great if we add comments to explain why the `card` prefix is needed here. It's an edge-case knowledge for almost every WCPay devs.**

**Comment by Jinksi** [comment]
> Update: it would be great if we add comments to explain why the card prefix is needed here. It's an edge-case knowledge for almost every WCPay devs.

Thanks @htdat! I've added the following comments in 8469698:
```
// Including the 'card' prefix to support subscription renewals using legacy payment method IDs.
```

**Comment by jessy-p** [comment]
Next step for this PR: Gamma to add tests before merging.

**Comment by naman03malhotra** [comment]
Adding 1sp for this effort. Jessy, please increase if required.

File: includes/core/server/request/class-create-intention.php [hunk 3]
```diff
@@ -47,7 +47,8 @@ public function get_method(): string {
 	 * @throws Invalid_Request_Parameter_Exception
 	 */
 	public function set_payment_method( string $payment_method_id ) {
-		$this->validate_stripe_id( $payment_method_id, [ 'pm', 'src' ] );
+		// Including the 'card' prefix to support subscription renewals using legacy payment method IDs.
+		$this->validate_stripe_id( $payment_method_id, [ 'pm', 'src', 'card' ] );
```
Comment (RadoslavGeorgiev) [inline includes/core/server/request/class-create-intention.php:51]: This is a different discussion, but I'm thinking of isolating those prefixes into a separate constants class. This way we'd have all definitions in one place, and a central explanation why `card` might be included in the mix.

Something like this:
```php
namespace WCPay\Core\Server;

abstract class Stripe_Prefix {
	/**
	 * Prefix for PaymentMethod objects.
	 *
	 * @var string
	 */
	const PAYMENT_METHOD = 'pm';

	/**
	 * Prefix for Source objects.
	 *
	 * Sources preceeded payment methods, and were never implemented in the
	 * gateway, but legacy stores should support them.
	 *
	 * @var string
	 */
	const SOURCE = 'src';

	/**
	 * Prefix for Card objects.
	 *
	 * Cards preceeded sources, and were never implemented in
	 * the gateway, but legacy stores should support them.
	 *
	 * @var string
	 */
	const CARD = 'card';
}
```

How does that sound?

**Review by RadoslavGeorgiev — COMMENTED** [review]
Thanks for taking this up @Jinksi! 🙌 

I left one comment, I'm curious whether it makes sense to emphasize on prefixes a bit more.

By the way, it's interesting. I seem to have added `src` [here](https://github.com/Automattic/woocommerce-payments/commit/ef10ff85f66961312a7c1ba7bdfd93dc46dd7668#diff-6a9840ad4b37c00ddd84090575c6cdbf2f6c1d6aa36ffa2a27e3413bb58b3e92R100), but never knew of needing to add `card` as well. We need to check whether that's needed for refunds as well in [this PR](https://github.com/Automattic/woocommerce-payments/pull/4663).

File: includes/core/server/request/class-create-intention.php (hunk 3 above)
Comment (htdat) [reply includes/core/server/request/class-create-intention.php:51]: > This is a different discussion, but I'm thinking of isolating those prefixes into a separate constants class. 

Good idea! But I think we don't often to use these values separately, so we just need to have an array aggregating those values like this:

```php
	const VALID_PAYMENT_METHODS = [ 'pm', 'src', 'card' ]; 
```

Comment (RadoslavGeorgiev) [reply includes/core/server/request/class-create-intention.php:51]: That would probably be even more useful, as we would not create a new request later, and miss any of them.  And with a PHPDoc comment, describing everything, that would be ideal.

**Comment by RadoslavGeorgiev** [comment]
@jessy-p the tests look great, thank you for adding them! 🙇  🟢

**Review by htdat — APPROVED** [review]
@jessy-p Pre-approve PR. Thanks for adding the new test file. 

Can we also add tests for two prefixes `src` and `card` in `tests/unit/core/server/request/test-class-core-create-and-confirm-intention-request.php` too? Currently, it's only testing `pm`.

**Review by htdat — APPROVED** [review]
@jessy-p - LGTM! 🚢 Thanks for adding more tests and addressing my feedback.
